File: baekjoon/src/tags/stack/No1874.java
package tags.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * https://www.acmicpc.net/problem/1874
 */
public class No1874 {
    public static final Stack<Integer> stack = new Stack<>();

    public static int[] array;

    public static List<Character> pushOrPop = new ArrayList<>();

    public static boolean isStackAvailable() {
        int numberToStack = 1;
        for (int value : array) {
            if (numberToStack <= value) {
                while (numberToStack <= value) {
                    stack.push(numberToStack++);
                    pushOrPop.add('+');
                }
            } else if (stack.peek() != value)
                return false;
            stack.pop();
            pushOrPop.add('-');
        }
        return true;
    }

    public static void main(String[] args) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            int numberOfInput = Integer.parseInt(reader.readLine());
            array = new int[numberOfInput];
            for (int i = 0; i < numberOfInput; i++)
                array[i] = Integer.parseInt(reader.readLine());
            if (isStackAvailable()) {
                for (char pushOrPop : pushOrPop)
                    System.out.println(pushOrPop);
            } else
                System.out.println("NO");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: baekjoon/src/tags/data_structures/No1874.kt
package tags.data_structures

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.Stack

/**
 * https://www.acmicpc.net/problem/1874
 */
private val stack = Stack<Int>()
private val pushOrPops = arrayListOf<Char>()

fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    val numberOfInput = reader.readLine().toInt()
    val numbers = IntArray(numberOfInput) { reader.readLine().toInt() }

    var maxNumber = 0
    for (number in numbers) {
        if (maxNumber < number) {
            for (i in maxNumber + 1..number)
                push(i)
            maxNumber = number
        } else {
            if (stack.peek() != number)
                break
        }
        pop()
    }

    if (stack.isEmpty()) {
        for (pushOrPop in pushOrPops) {
            writer.write("$pushOrPop")
            writer.newLine()
        }
    } else
        writer.write("NO")
    writer.flush()
}

private fun push(number: Int) {
    stack.push(number)
    pushOrPops.add('+')
}

private fun pop() {
    stack.pop()
    pushOrPops.add('-')
}
